File: test/client/common/gateways/GitHubGatewayTest.js
import {withMockedImports} from '../../TestUtils'
import {describe, it} from 'mocha'
import {expect} from 'chai'
import {mocks} from '../../Mocking'
import {GatewayError} from '../../../../src/client/common/gateways/Gateway'

describe('GitHubGateway', function () {

  const EXPECTED_ACCEPT_HEADER = {Accept: 'application/vnd.github.v3+json'}

  const get = mocks.stub()
  const patch = mocks.stub()
  const post = mocks.stub()
  const gatewaySend = mocks.stub()

  const {createGist, updateGist, getGist, getTruncatedFile, send} = withMockedImports('client/common/gateways/GitHubGateway', {
    './Gateway': {send: gatewaySend, get, patch, post}
  })

  describe('createGist', function () {

    it('should call with the gist URL, updated gist data and the correct headers', function () {
      const expectedData = {
        description: 'some-description',
        public: false,
        files: {'configuration.json': {content: 'some-configuration'}}
      }
      const expectedHeaders = {
        Authorization: 'token some-oauth-token',
        ...EXPECTED_ACCEPT_HEADER
      }

      createGist('some-description', 'some-configuration', 'some-oauth-token')

      expect(post).to.have.been.calledWith('https://api.github.com/gists', expectedData, expectedHeaders)
    })
  })

  describe('updateGist', function () {

    it('should call with the gist URL, updated gist data and the correct headers', function () {
      const expectedData = {
        description: 'some-description',
        files: {'configuration.json': {content: 'some-configuration'}}
      }
      const expectedHeaders = {
        Authorization: 'token some-oauth-token',
        ...EXPECTED_ACCEPT_HEADER
      }

      updateGist('some-gist-id', 'some-description', 'some-configuration', 'some-oauth-token')

      expect(patch).to.have.been.calledWith('https://api.github.com/gists/some-gist-id', expectedData, expectedHeaders)
    })
  })

  describe('getGist', function () {

    it('should call with the gist URL and the correct accept header', function () {
      getGist('some-gist-id')
      expect(get).to.have.been.calledWith('https://api.github.com/gists/some-gist-id', EXPECTED_ACCEPT_HEADER)
    })
  })

  describe('getTruncatedFile', function () {

    it('should call with the given URL and the correct accept header', function () {
      getTruncatedFile('some-url')
      expect(get).to.have.been.calledWith('some-url', {Accept: 'text/plain; charset=utf-8'})
    })
  })

  describe('send', function () {

    it('should return the message from the body including the status on error', async function () {
      gatewaySend.rejects(new GatewayError({status: 500, body: {message: 'some-error'}}))
      try {
        await send()
      } catch (err) {
        expect(err).to.have.property('message', '500 - some-error')
      }
    })

    it('should return the body if it does not contain a message on error', async function () {
      gatewaySend.rejects(new GatewayError({status: 0, body: 'timeout'}))
      try {
        await send()
      } catch (err) {
        expect(err).to.have.property('message', '0 - timeout')
      }
    })

    it('should return the given status on error', async function () {
      gatewaySend.rejects(new GatewayError({status: 500}))
      try {
        await send()
      } catch (err) {
        expect(err).to.have.property('status', 500)
      }
    })
  })
})

File: src/server/nevergreen/backup/github_gateway_test.clj
(ns nevergreen.backup.github-gateway-test
  (:require [clojure.test :refer :all]
            [nevergreen.backup.github-gateway :as subject]))

(deftest create-gist

  (testing "allows the URL to be set to support GitHub Enterprise"
    (binding [subject/http-post (fn [url _]
                                  (is (= "https://some-host/gists" url)))]
      (subject/create-gist {:description "" :configuration "" :token "" :url "https://some-host"})
      (subject/create-gist {:description "" :configuration "" :token "" :url "https://some-host/"})))

  (testing "uses the default GitHub API URL if none provided"
    (binding [subject/http-post (fn [url _]
                                  (is (= "https://api.github.com/gists" url)))]
      (subject/create-gist {:description "" :configuration "" :token "" :url nil})
      (subject/create-gist {:description "" :configuration "" :token "" :url ""})))

  (testing "sets the correct body"
    (binding [subject/http-post (fn [_ data]
                                  (is (= "{\"description\":\"some-description\",\"public\":false,\"files\":{\"configuration.json\":{\"content\":\"some-config\"}}}"
                                         (:body data))))]
      (subject/create-gist {:description "some-description" :configuration "some-config" :token "some-token"})))

  (testing "sets the correct mime type"
    (binding [subject/http-post (fn [_ data]
                                  (is (= "application/vnd.github.v3+json"
                                         (:accept data))))]
      (subject/create-gist {:description "some-description" :configuration "some-config" :token "some-token"})))

  (testing "sets the correct auth header"
    (binding [subject/http-post (fn [_ data]
                                  (is (= {"Authorization" "token some-token"}
                                         (:headers data))))]
      (subject/create-gist {:description "some-description" :configuration "some-config" :token "some-token"}))))

(deftest update-gist

  (testing "allows the URL to be set to support GitHub Enterprise"
    (binding [subject/http-patch (fn [url _]
                                   (is (= "https://some-host/gists/some-id" url)))]
      (subject/update-gist {:description "" :configuration "" :token "" :id "some-id" :url "https://some-host"})
      (subject/update-gist {:description "" :configuration "" :token "" :id "some-id" :url "https://some-host/"})))

  (testing "uses the default GitHub API URL if none provided"
    (binding [subject/http-patch (fn [url _]
                                   (is (= "https://api.github.com/gists/some-id" url)))]
      (subject/update-gist {:description "" :configuration "" :token "" :id "some-id" :url nil})
      (subject/update-gist {:description "" :configuration "" :token "" :id "some-id" :url ""})))

  (testing "sets the correct body"
    (binding [subject/http-patch (fn [_ data]
                                   (is (= "{\"description\":\"some-description\",\"public\":false,\"files\":{\"configuration.json\":{\"content\":\"some-config\"}}}"
                                          (:body data))))]
      (subject/update-gist {:description "some-description" :configuration "some-config" :token "some-token" :id "some-id"})))

  (testing "sets the correct mime type"
    (binding [subject/http-patch (fn [_ data]
                                   (is (= "application/vnd.github.v3+json"
                                          (:accept data))))]
      (subject/update-gist {:description "some-description" :configuration "some-config" :token "some-token" :id "some-id"})))

  (testing "sets the correct auth header"
    (binding [subject/http-patch (fn [_ data]
                                   (is (= {"Authorization" "token some-token"}
                                          (:headers data))))]
      (subject/update-gist {:description "some-description" :configuration "some-config" :token "some-token" :id "some-id"}))))

(deftest get-gist

  (testing "allows the URL to be set to support GitHub Enterprise"
    (binding [subject/http-get (fn [url _]
                                 (is (= "https://some-host/gists/some-id" url)))]
      (subject/get-gist "some-id" "https://some-host")
      (subject/get-gist "some-id" "https://some-host/")))

  (testing "allows the URL to be set to support GitHub Enterprise"
    (binding [subject/http-get (fn [url _]
                                 (is (= "https://api.github.com/gists/some-id" url)))]
      (subject/get-gist "some-id" nil)
      (subject/get-gist "some-id" "")))

  (testing "sets the correct mime type"
    (binding [subject/http-get (fn [_ data]
                                 (is (= "application/vnd.github.v3+json"
                                        (:accept data))))]
      (subject/get-gist "some-id" nil))))

(deftest get-truncated-file

  (testing "uses the given URL (as it comes from getting the gist)"
    (binding [subject/http-get (fn [url _]
                                 (is (= "some-url" url)))]
      (subject/get-truncated-file "some-url")))

  (testing "sets the correct mime type"
    (binding [subject/http-get (fn [_ data]
                                 (is (= "text/plain; charset=utf-8"
                                        (:accept data))))]
      (subject/get-truncated-file "some-url")))

  (testing "gets the response as a string as the server shouldn't know what format the data is"
    (binding [subject/http-get (fn [_ data]
                                 (is (nil? (:as data))))]
      (subject/get-truncated-file "some-url"))))
